main.go

```diff
@@ -95,114 +101,171 @@ func getEtcdConfig() (etcd.Config, error) {
 	return config, nil
 }
 
-func getGlobalASN(api etcd.KeysAPI) (uint32, error) {
-	res, err := api.Get(context.Background(), fmt.Sprintf("%s/global/as_num", CALICO_BGP), nil)
+type Server struct {
+	t         tomb.Tomb
+	bgpServer *bgpserver.BgpServer
+	cli       *calicocli.Client
+	etcd      etcd.KeysAPI
+}
+
+func NewServer() (*Server, error) {
+	config, err := calicocli.LoadClientConfigFromEnvironment()
 	if err != nil {
-		return 0, err
+		return nil, err
 	}
-	asn, err := strconv.ParseUint(res.Node.Value, 10, 32)
+
+	etcdConfig, err := getEtcdConfig(config)
 	if err != nil {
-		return 0, err
+		return nil, err
+	}
+
+	cli, err := etcd.New(etcdConfig)
+	if err != nil {
+		return nil, err
+	}
+	etcdCli := etcd.NewKeysAPI(cli)
+
+	calicoCli, err := calicocli.New(*config)
+	if err != nil {
+		return nil, err
+	}
+
+	bgpServer := bgpserver.NewBgpServer()
+
+	return &Server{
+		bgpServer: bgpServer,
+		cli:       calicoCli,
+		etcd:      etcdCli,
+	}, nil
+}
+
+func (s *Server) Serve() {
+	s.t.Go(func() error {
+		s.bgpServer.Serve()
+		return nil
+	})
+
+	bgpAPIServer := bgpapi.NewGrpcServer(s.bgpServer, ":50051")
+	s.t.Go(bgpAPIServer.Serve)
+
+	globalConfig, err := s.getGlobalConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := s.bgpServer.Start(globalConfig); err != nil {
+		log.Fatal(err)
 	}
-	return uint32(asn), nil
+
+	if err := s.initialPolicySetting(); err != nil {
+		log.Fatal(err)
+	}
+
+	// monitor routes from other BGP peers and update FIB
+	s.t.Go(s.monitorPath)
```

Comment:
Does the difference between monitorX and watchX refer to something subtly different, or could they all be watchX?